src/unitxt/operators.py

```diff
@@ -213,6 +213,103 @@ def process(
         return flatten_dict(instance, parent_key=self.parent_key, sep=self.sep)
 
 
+class ModelInputFormatter(StreamInstanceOperator):
+    r"""Generates the whole input to the model, from constant strings that are given as args, and from values found in specified fields of the instance.
```

Comment:
I think the docstring should be simplified and focused on the required inputs and produces outputs. Other things such as description of the works of other parts of unitxt should be ommited as much as possible, IMO. Here is my sketchy attempt, please consider something along these lines:

ModelInputFormatter expects the input instance to contain: 
    1. A "source" field with the source dataset instance values represented by a single string, in the context of the underlying task. 
    2. A "instruction" field containing a non-None value.
    3. A field named with the value in `self.demos_field`, containing a list of dicts, each dict with "source" and "target" field representing a single demo.
    4. A field named with the value in `self.system_prompt_field`
    
ModelInputFormatter formats the above fields into a single string inputted to the model. This string overwrites the "source" field of the instance. 
ModelInputFormatter also pops fields "instruction" and "demos" from the input instance.

Args:
    demos_field (str): the name of the field that contains the demos, being dicts with "source" and "target" keys
    system_prompt_field (str): the name of the field containing an introductory text and or xml tags
    demo_format (str): formatting string for a single demo, combining fields source and target
    model_input_format (str) overall output format, combining system_prompt, instruction, demos, and source

Reply:
Thanks @matanor , I polished the docstring as you elegantly suggested.